x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java

```diff
@@ -709,6 +848,12 @@ private synchronized boolean addSetStopAtCheckpointListener(
                     )
                 );
             }
+
+            // stop the transform if the decision was to stop it above
+            if (newtaskState.equals(TransformTaskState.STOPPED)) {
+                context.shutdown();
```

Comment:
the shutdown now happens in the thread that locked the task and the indexer. shutdown locks the task once more.